<a href="https://colab.research.google.com/github/ravipm260372/AI4AstroMasterclass/blob/master/D_Phi_Introduction_to_Pandas_08_Mar_2020.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 1
# `Pandas`, Data Analysis library


`Pandas` is a data analysis library that provides a variety of data structures and data manipulation methods that allows to perform complex tasks with simple one-line commands.

In the following sections, we will be working on the `Pandas` library and its uses. We will be reviewing how the objects are defined by `pandas`, operating their values and executing a series of important operations for data analysis. Let us begin!

## 1. Installing and Using Pandas

The downloading and installation of `pandas` library process can be done through the `pip` standard package-management system, by executing in your local environment's console this single line:

`pip install pandas`

Once installed, we can import it in the following way:

In [ ]:
import pandas
pandas.__version__

'1.1.5'

It is good practice to create the alias `pd` for `pandas`:

In [ ]:
import pandas as pd

## 2. Reviewing Pandas Objects

At the core of the `` pandas`` library there are two fundamental data structures/objects:
1. **`Series`**: stores single column data along with an **index**. An index is just a way to "number" the `Series` object.
2. **`DataFrame`**: is a two-dimensional tabular data structure with labeled axes. It is conceptually useful to think of a `DataFrame` object as a collection of `Series` objects. That is, think of each column in a DataFrame as a single object in the `Series`, where each of these objects in the `Series` shares a common index: the index of the `DataFrame` object.

Let's import the necessary libraries and we will delve into each of these `pandas` concepts:

In [ ]:
import numpy as np
import pandas as pd

### 2.1. Pandas `Series` Object



A Pandas ``Series`` is a one-dimensional array of indexed data.
It can be created from a list or array as follows:

In [ ]:
data = pd.Series([0.25, 0.5, 0.75, 1.0])
data

0    0.25
1    0.50
2    0.75
3    1.00
dtype: float64

As we see in the output, the `Series` wraps both a sequence of values and a sequence of indices starting from 0 to the number of values added through the list, which we can access with the `values` and `index` attributes. The values are simply a familiar `NumPy` array:

In [ ]:
data.values

array([0.25, 0.5 , 0.75, 1.  ])

The index is an array-like object of type `pd.Index`.

In [ ]:
data.index

RangeIndex(start=0, stop=4, step=1)

Like with a NumPy array, data can be accessed by the associated index via the familiar Python square-bracket slicing notation:

In [ ]:
data[1]

0.5

In [ ]:
data[1:3]

1    0.50
2    0.75
dtype: float64

As we will see, the Pandas ``Series`` is much more general and flexible than the one-dimensional NumPy array that it emulates.

#### 2.1.1. `Series` as generalized NumPy array

`Numpy` Array has an implicitly defined integer index used to access the values. The Pandas `Series` has an explicitly defined `index` associated with the values.

This explicit `index` definition gives the `Series` object additional capabilities. The `index` does not need an integer value mandatory, but can consist of values of any desired type, as we can see in the following example:

In [ ]:
data = pd.Series([0.25, 0.5, 0.75, 1.0],
                 index=['a', 'b', 'c', 'd'])
data

a    0.25
b    0.50
c    0.75
d    1.00
dtype: float64

As we can see, we have defined the indices as characters of the English alphabet. One important fact is that the number of defined indices corresponds to the number of added values in the `Series`. When accessing the values associated with the indexes, we must consider the new index nomenclature and access the values with the common slicing:

In [ ]:
data['b']

0.5

We can even use non-contiguous or non-sequential indices (index), considering the number of defined indices corresponds to the number of added values in the `Series`:

In [ ]:
data = pd.Series([0.25, 0.5, 0.75, 1.0],
                 index=[2, 5, 3, 7])
data

2    0.25
5    0.50
3    0.75
7    1.00
dtype: float64

And we access the values by slicing the indices such as:

In [ ]:
data[5]

0.5

#### 2.1.2 `Series` as specialized dictionaries

A dictionary is a structure that maps arbitrary keys to a set of arbitrary values, and a `Series` is a structure which maps typed keys to a set of typed values. We can take advantage of these similarities to create a `Series` from a dictionary, where the `keys` are the `indices` of the `Series` and the `values` are those associated with these `indices`.

In [ ]:
population_dict = {'California': 38332521,
                   'Texas': 26448193,
                   'New York': 19651127,
                   'Florida': 19552860,
                   'Illinois': 12882135}
population = pd.Series(population_dict)
population

California    38332521
Texas         26448193
New York      19651127
Florida       19552860
Illinois      12882135
dtype: int64

From here, typical dictionary-style item access can be performed:

In [ ]:
population['California']

38332521

Unlike a dictionary, the ``Series`` also supports array-style operations such as slicing:

In [ ]:
population['California':'Illinois']

California    38332521
Texas         26448193
New York      19651127
Florida       19552860
Illinois      12882135
dtype: int64

### 2.2. Pandas `DataFrame` Object

A ``DataFrame`` is an analog of a two-dimensional array with both flexible row indices and flexible column names.
Just as you might think of a two-dimensional array as an ordered sequence of aligned one-dimensional columns, you can think of a ``DataFrame`` as a sequence of aligned ``Series`` objects.
Here, by "aligned" we mean that they share the same index.

To demonstrate this, let's first construct a new ``Series`` listing the area of each of the five states discussed in the previous section:

In [ ]:
area_dict = {'California': 423967, 'Texas': 695662, 'New York': 141297,
             'Florida': 170312, 'Illinois': 149995}
area = pd.Series(area_dict)
area

California    423967
Texas         695662
New York      141297
Florida       170312
Illinois      149995
dtype: int64

Now that we have this along with the ``population`` Series from before, we can use a dictionary to construct a single two-dimensional object containing this information:

In [ ]:
states = pd.DataFrame({'population': population,
                       'area': area})
states

population    area
California    38332521  423967
Texas         26448193  695662
New York      19651127  141297
Florida       19552860  170312
Illinois      12882135  149995

Like the ``Series`` object, the ``DataFrame`` has an ``index`` attribute that gives access to the index labels:

In [ ]:
states.index

Index(['California', 'Texas', 'New York', 'Florida', 'Illinois'], dtype='object')

Additionally, the ``DataFrame`` has a ``columns`` attribute, which is an ``Index`` object holding the column labels:

In [ ]:
states.columns

Index(['population', 'area'], dtype='object')

Thus the ``DataFrame`` can be thought of as a generalization of a two-dimensional NumPy array, where both the rows and columns have a generalized index for accessing the data.

There are some functions and attributes that allow us to observe basic information about the data stored in a `DataFrame` object:

1. `DataFrame.head()` -> returns the content of the first 5 rows, by default
2. `DataFrame.tail()` -> returns the content of the last 5 rows, by default
3. `DataFrame.shape` -> returns a tuple of the form (num_rows, num_columns)
4. `DataFrame.columns` -> returns the name of the columns
5. `DataFrame.index` -> returns the index of the rows

Using `` `data.head ()` `` and `` `data.tail ()` `` we can see the content of the data. Unless otherwise specified, `DataFrame` and `Series` objects have indexes starting from 0 and incrementing monotonically and incrementally as integers.

In [ ]:
states.head(3) # The first three rows

population    area
California    38332521  423967
Texas         26448193  695662
New York      19651127  141297

In [ ]:
states.tail(3) # The last three rows

population    area
New York    19651127  141297
Florida     19552860  170312
Illinois    12882135  149995

### 2.3. Pandas `Index` Object

This ``Index`` object is an interesting structure itself, and it can be thought as an *immutable array*:

In [ ]:
ind = pd.Index([2, 3, 5, 7, 11])
ind

Int64Index([2, 3, 5, 7, 11], dtype='int64')

This index object can be sliced as a list or a numpy array:

In [ ]:
ind[1] # Accessing element in position 1

3

In [ ]:
ind[::2] # Accessing elements starting from position 0 through all the elements two by two.

Int64Index([2, 5, 11], dtype='int64')

Some common attributes with `Numpy` arrays:

In [ ]:
print(' Size:',ind.size,'\n',
      'Shape:',ind.shape,'\n',
      'Dimension:',ind.ndim,'\n', 
      'Data type:',ind.dtype)

 Size: 5 
 Shape: (5,) 
 Dimension: 1 
 Data type: int64


One difference between ``Index`` objects and `Numpy` arrays is that indices are **immutable**. That is, they cannot be modified via the normal means, which will cause an error:

In [ ]:
ind[1] = 0

TypeError: ignored

## 3. Data Indexing and Selection

Let's see in detail how to access the elements of the `Series` and `DataFrames` objects.

### 3.1. Data Selection in `Series`

Let's redefine a `Series` object for explanatory purposes:

In [ ]:
data = pd.Series([0.25, 0.5, 0.75, 1.0],
                 index=['a', 'b', 'c', 'd'])
data

a    0.25
b    0.50
c    0.75
d    1.00
dtype: float64

In addition to accessing the elements of a `Series` through slicing the indexes, we can update the values associated with them as:

In [ ]:
data['d'] = 0.95 # updating the value associated with 'd' index in Series object
data

a    0.25
b    0.50
c    0.75
d    0.95
dtype: float64

We can also add a new value using the same procedure, just as we would do with a dictionary:

In [ ]:
data['e'] = 1.25 # New value added with 'e' index in Series object
data

a    Hello
b     DPhi
c    world
e     1.25
dtype: object

Slicing access can be done *explicitly* or *implicitly* through a range, such that:

In [ ]:
# slicing by explicit index
data['a':'c']

a    Hello
b     DPhi
c    world
dtype: object

In [ ]:
# slicing by implicit integer index
data[0:2]

a    0.25
b    0.50
dtype: float64

We can access a group of specific indices by including these indices in a list, in such a way that access is done as follows:

Notice that when slicing with an explicit index (i.e., `data ['a':'c']`), the final index is included in the slice, while when slicing with an implicit index (i.e., `data[0:2]`), the final index is excluded from the slice. When we slicing through a list (i.e., `data [['a', 'e']]`), all indices are equally accessed.

### 3.2. Indexers: loc, iloc for `Series`

We are going to review two ways to access the elements of a `Series`, using two attributes determined for this: `.loc[]` and `.iloc[]`. Let's first define a `Series` of three string elements and character indices:

In [ ]:
data = pd.Series(['Hello', 'DPhi', 'world'], index=['a', 'b', 'c'])
data

a    Hello
b     DPhi
c    world
dtype: object

#### 3.2.1. `loc` attribute

The ``loc`` attribute allows indexing and slicing that always references the explicit index (the explicit name of the index):

In [ ]:
data.loc['a']

0.25

In [ ]:
data.loc['a':'c']

#### 3.2.2. `iloc` attribute

The ``iloc`` attribute allows indexing and slicing that always references the implicit Python-style index (the direct row number of the values monotonically increasing from 0 to 3):

In [ ]:
data.iloc[1]

In [ ]:
data.iloc[1:3]

## Exercise 1

Consider the following lists:
```
lst1 = [1, 2, 3, 5, 8]
lst2 = [8, 5, 3, 2, 1]
```

1. Create and display two individual `Series` objects `s1` and `s2` from the data available on each list.


2. Perform the following operations with the two series (element-wise):
    1. Add `s1` and `s2` and store the result in a new variable `s3_add`
    2. Subtract `s2` from `s1` and store the result in a new variable `s3_sub`
    3. Multiply `s1` and `s2` and store the result in a new variable `s3_mul`
    4. Divide `s1` by `s2` and store the result in a new variable `s3_div`

In [ ]:
## Answer to Question 1

In [ ]:
# create the series objects
s1 = pd.Series([1,2,3,5,8])
s2 = pd.Series([8,5,3,2,1])

In [ ]:
# display 1st series
s1

0    1
1    2
2    3
3    5
4    8
dtype: int64

In [ ]:
# display 2nd series
s2

0    8
1    5
2    3
3    2
4    1
dtype: int64

In [ ]:
# Answer to Question 2 : 1st part

In [ ]:
# add the two series and store in s3_add
s3_add = s1 + s2
s3_add

0    9
1    7
2    6
3    7
4    9
dtype: int64

In [ ]:
# Answer to Question 2 : 2nd part

In [ ]:
# Subtract s2 from s1 and store the result in a new variable s3_sub
s3_sub = s1-s2
s3_sub

0   -7
1   -3
2    0
3    3
4    7
dtype: int64

In [ ]:
# Answer to Question 2 : 3rd part

In [ ]:
# Multiply s1 and s2 and store the result in a new variable s3_mul
s3_mul = s1 * s2
s3_mul

0     8
1    10
2     9
3    10
4     8
dtype: int64

In [ ]:
# Answer to Question 2 : 4tf part

In [ ]:
# Divide s1 by s2 and store the result in a new variable s3_div
s3_div = s1/s2
s3_div

0    0.125
1    0.400
2    1.000
3    2.500
4    8.000
dtype: float64

## Exercise 2

Consider the following `Series` object:
```
0    45000
1    37872
2    57923
3    68979
4    78934
5    69897
6    56701
Name: Amazon_Reviews, dtype: int64
```

1. Create and display the `Amazon_Reviews` Series.

2. Get the last three values from `Amazon_Reviews` using negative indexing.

In [ ]:
# Answer 1 : Create and display the Amazon_Reviews Series
Amazon_Reviews = pd.Series([45000 , 37872 , 57923 , 68979 , 78934 , 69897 , 56701])
Amazon_Reviews

0    45000
1    37872
2    57923
3    68979
4    78934
5    69897
6    56701
dtype: int64

In [ ]:
# Answer 2 Get the last three values from Amazon_Reviews using negative indexing.
Amazon_Reviews[-3:]

4    78934
5    69897
6    56701
dtype: int64

## Exercise 3

Consider the following dictionary which is relating the area in sq units of some USA states: 
```
    area_dict = {'California': 423967, 'Texas': 695662, 'New York': 141297,
             'Florida': 170312, 'Illinois': 149995}
```

1. Create a `Series` using the given dictionary
2. Extract areas for 'Texas', 'New York',  and 'Florida' from the created series

In [ ]:
# Answer 1 : Create a Series using the given dictionary area_dict
# create the area_dict dictionary
area_dict = {'California': 423967, 'Texas': 695662, 'New York': 141297,
             'Florida': 170312, 'Illinois': 149995}

# create the series area_dict
area_dict = pd.Series(area_dict)

# print the series
area_dict

California    423967
Texas         695662
New York      141297
Florida       170312
Illinois      149995
dtype: int64

In [ ]:
# Answer 2 : Extract areas for 'Texas', 'New York', and 'Florida' from the created series
# extract series elemnts using clicing
area_dict['Texas' : 'Florida']

Texas       695662
New York    141297
Florida     170312
dtype: int64

### 3.3. Data Selection in `DataFrame`

Let's see in detail how to access the elements of the `DataFrame` objects, redefine a `DataFrame` object for explanatory purposes:

In [ ]:
area = pd.Series({'California': 423967, 'Texas': 695662,
                  'New York': 141297, 'Florida': 170312,
                  'Illinois': 149995})
pop = pd.Series({'California': 38332521, 'Texas': 26448193,
                 'New York': 19651127, 'Florida': 19552860,
                 'Illinois': 12882135})
data = pd.DataFrame({'area':area, 'pop':pop})
data

area       pop
California  423967  38332521
Texas       695662  26448193
New York    141297  19651127
Florida     170312  19552860
Illinois    149995  12882135

The individual ``Series`` that make up the columns of the ``DataFrame`` can be accessed via dictionary-style indexing of the column name:

In [ ]:
data['area'] # Accessing the 'area' column of the data DataFrame

California    423967
Texas         695662
New York      141297
Florida       170312
Illinois      149995
Name: area, dtype: int64

In [ ]:
data['pop'] # Accessing the 'pop' column of the data DataFrame

California    38332521
Texas         26448193
New York      19651127
Florida       19552860
Illinois      12882135
Name: pop, dtype: int64

Equivalently, we can use attribute-style access with column names that are strings, which will result in the exact same `Series` output:

In [ ]:
data.area # Equivalent to data['area']

California    423967
Texas         695662
New York      141297
Florida       170312
Illinois      149995
Name: area, dtype: int64

Like with the ``Series`` objects discussed earlier, this dictionary-style syntax can also be used to modify the object, in this case adding a new column:

In [ ]:
data['density'] = data['pop']/data['area']
data

area       pop     density
California  423967  38332521   90.413926
Texas       695662  26448193   38.018740
New York    141297  19651127  139.076746
Florida     170312  19552860  114.806121
Illinois    149995  12882135   85.883763

As you can see, when the `Series` 'pop' is accessed and divided over the `Series` 'area', the arithmetic operation becomes element-wise and the result is assigned to the new `Series` 'density', which becomes the third column of the `DataFrame` `data`.

We can also view the ``DataFrame`` as an enhanced two-dimensional array.
We can examine the raw underlying data array using the ``values`` attribute, which will return a two-dimensional array in which each row corresponds to a row of `DataFrame` values:

In [ ]:
data.values

array([[4.23967000e+05, 3.83325210e+07, 9.04139261e+01],
       [6.95662000e+05, 2.64481930e+07, 3.80187404e+01],
       [1.41297000e+05, 1.96511270e+07, 1.39076746e+02],
       [1.70312000e+05, 1.95528600e+07, 1.14806121e+02],
       [1.49995000e+05, 1.28821350e+07, 8.58837628e+01]])

### 3.4. Indexers: loc, iloc for `DataFrame`

We are going to review two ways to access the elements of a `DataFrame`, using two attributes determined for this: `.loc[]` and `.iloc[]`:

#### 3.4.1. `loc` attribute

The ``loc`` attribute allows indexing and slicing that always references the explicit index (the explicit name of the index):

In [ ]:
data.loc[:'Illinois', :'pop']

area       pop
California  423967  38332521
Texas       695662  26448193
New York    141297  19651127
Florida     170312  19552860
Illinois    149995  12882135

In this example we are slicing on the indices starting with the first by default and ending in 'Illinois', as well as by the columns starting with the first by default, and ending in 'pop'.

#### 3.4.2. `iloc` attribute

The ``iloc`` attribute allows indexing and slicing that always references the implicit Python-style index (the direct row number of the values monotonically increasing from 0 to 3):

In [ ]:
data.iloc[:3, :2]

area       pop
California  423967  38332521
Texas       695662  26448193
New York    141297  19651127

In this example we are slicing the `DataFrame` from index 0 by default to 3 exlusive, and from column 0 by default to 2 exlusive.

## Exercise 4

Consider below DPhi Bootcamp's information about different batches:

```
Total_Candidates = {'absolute_beginners': 785, 'beginners': 825, 'intermediat_advanced': 602} # this is true data
Active_Candidates = {'absolute_beginners': 500, 'beginners': 425, 'intermediat_advanced': 300}  # this is hypothetical data
```
    
1. Create a Pandas `DataFrame` using above information (name your Dataframe as `DPhi`)
2. Get all the columns in DPhi.
3. Get the information of total candidates present in each batches using dictionary-style indexing.
4. Find the number of candidates for each batches who are not active and add this information to the dataframe DPhi.
5. Also, find the percent of candidates that are active in each batches and add this information to the `DPhi` dataframe (hint: $percent = (active / total)* 100$)
6. Get all the batches where percentage of active candidates are greater than 60%

In [ ]:
# Answer 1 : Create a Pandas DataFrame DPhi

# create the dictionaries of Total_Candidates and Active_candidates
Total_Candidates = {'absolute_beginners': 785, 'beginners': 825, 'intermediat_advanced': 602} # this is true data
Active_Candidates = {'absolute_beginners': 500, 'beginners': 425, 'intermediat_advanced': 300}  # this is hypothetical data

# create the DatFrame DPhi using the pd.DataFrame method
DPhi = pd.DataFrame({'Total_Candidates': Total_Candidates,
                       'Active_Candidates': Active_Candidates})

# view the DataFrame
DPhi

Total_Candidates  Active_Candidates
absolute_beginners                 785                500
beginners                          825                425
intermediat_advanced               602                300

In [ ]:
# Answer 2 : Get all the columns in DPhi.
DPhi.columns

Index(['Total_Candidates', 'Active_Candidates'], dtype='object')

In [ ]:
# Answer 3 : Get the information of total candidates present in each batches using dictionary-style indexing.
DPhi.iloc[:3 , :3]

Total_Candidates  Active_Candidates
absolute_beginners                 785                500
beginners                          825                425
intermediat_advanced               602                300

In [ ]:
# Answer 4 : Find the number of candidates for each batches who are not active and add this information to the dataframe DPhi.
# create a new column Not_Active which is difference of the two columns
DPhi['Not_Active'] = DPhi['Total_Candidates'] - DPhi['Active_Candidates']
DPhi

Total_Candidates  Active_Candidates  Not_Active
absolute_beginners                 785                500         285
beginners                          825                425         400
intermediat_advanced               602                300         302

In [ ]:
# Answer 5 : Find the percent of candidates that are active in each batches and add this information to the DPhi dataframe (hint:  percent=(active/total)∗100 )
# create a new column "Percent_Active = (Active_Candidates / Total_Candidates)*100"
DPhi['Percent_Active'] = np.round(DPhi['Active_Candidates']/DPhi['Total_Candidates']*100 , 2)
DPhi

Total_Candidates  ...  Percent_Active
absolute_beginners                 785  ...           63.69
beginners                          825  ...           51.52
intermediat_advanced               602  ...           49.83

[3 rows x 4 columns]

In [ ]:
# Answer 6 : Get all the batches where percentage of active candidates are greater than 60%
# use comparison operator to slice the DataFrame where Percent_Active > 60
grt_than_60 = DPhi[DPhi['Percent_Active'] > 60]
grt_than_60

Total_Candidates  ...  Percent_Active
absolute_beginners               785  ...           63.69

[1 rows x 4 columns]

### 3.5. Subsetting a `Dataframe`

**Subsetting** a `DataFrame` is a way of filtering which allows to extract portions of interest. Subsetting can be done using comparison operators and logical operators inside a pair of square brackets `[]` as shown in the following example:

In [ ]:
data[data['density'] > 100]

area       pop     density
New York  141297  19651127  139.076746
Florida   170312  19552860  114.806121

In the above example, we are extracting the rows for which it is `True` that the population density is greater than 100 units. For more clarification on the logical operation concept, take a look to the following extract of the above example:

In [ ]:
data['density'] > 100

California    False
Texas         False
New York       True
Florida        True
Illinois      False
Name: density, dtype: bool

Since there are only two rows where the density is greater than 100, the result will be a boolean-values `DataFrame` in which the `True` values correspond to the rows that accomplish the locigal expression, and the `False` values to the ones that do not.

Let's see an example in which we include a second logical operation to select those rows in which the population density is greater than 100 units `data['density'] > 100` and (`&`) the area is less than 150,000 units `data['area'] < 150000`:

In [ ]:
data[(data['density'] > 100) & (data['area'] < 150000)]

area       pop     density
New York  141297  19651127  139.076746

We could also select those records in which the population density is less than 90 units `data['density'] < 90` or (`|`) greater than 120 `data['density'] > 120`:

In [ ]:
data[(data['density'] < 90) | (data['density'] > 120)]

area       pop     density
Texas     695662  26448193   38.018740
New York  141297  19651127  139.076746
Illinois  149995  12882135   85.883763

The previous example can be rewritten to express the negation of the above condition, select those records that do not have a population density greater than or equal to 90 units and less than or equal to 120 units `~((data['density'] >= 90) & (data['density'] <= 120))`:

In [ ]:
data[~((data['density'] >= 90) & (data['density'] <= 120))]

area       pop     density
Texas     695662  26448193   38.018740
New York  141297  19651127  139.076746
Illinois  149995  12882135   85.883763

As we can see, the results of the last two examples are the same, since we are expressing the same condition in different ways.

## 4. Data Wrangling

The difference between data found in many tutorials and data from the real world is that real-world data is rarely clean and homogeneous. In particular, many interesting datasets will have some amount of data missing. To make matters even more complicated, different data sources may indicate missing data in different ways. 

In this way, we need to define methods that allow us to structure, clean and enrich the data acquired from the real world, which are the main steps for **Data Wrangling**. Before continuing, let's see what is the difference between these three steps and expand their definition:

**1. Data structuring:** 

The first step in the data wrangling process is to separate the relevant data into multiple columns, so that the analysis can be run by grouping common values in a separate way. In turn, if there are columns that are not desired or that will not be relevant to the analysis, this is the phase to filter the data or mix together some of their columns.

**2. Data Cleaning**

In this step, the data is cleaned up for high-quality analysis. `Null values` are handled, and the data format is standardized. We will enter this process in the following weeks.

**3. Data Enriching**

After cleaning, the data is enriched by increasing some variables in what is known as *Data Augmentation* and using additional sources to enrich them for the following stages of processing.

For now, we will review how to handle missing values, a fundamental step for data cleaning.

## 5. Handling Missing Data

This is a fundamental step in data cleaning. It is common that during the data acquisition processes, there are lost records, either due to the difficulties of acquiring them, due to errors in the source or destination, or because we simply could not acquire the data. There are three types of missing data:

- Missing completely at random (MCAR): when the fact that the data is missing is independent of the observed and unobserved data.
- Missing at random (MAR): when the fact that the data is missing is systematically related to the observed but not the unobserved data.
- Missing not at random (MNAR): when the missingness of data is related to events or factors which are not measured by the researcher.

We will go into these types in detail later. For now, we'll look at the fundamentals of handling missing data in pandas:

### 5.1. `NaN` and `None` in Pandas

Missing data is handled in Pandas as `NaN` values placeholders. `NaN` value is a IEEE 754 floating point representation of *Not a Number (NaN)*. One of the main reasons to handle missing data as `NaN` rather than `Null` in Pandas is that `NaN` (from `np.nan`) allows for vectorized operations, since it is a float value. `None`, by definition, forces object type, which basically disables all efficiency in Numpy and Pandas.

``NaN`` and ``None`` are handled nearly interchangeably by `Pandas`, converting between them where appropriate:

In [ ]:
import numpy as np

pd.Series([1, np.nan, 2, None])

0    1.0
1    NaN
2    2.0
3    NaN
dtype: float64

### 5.2. Operations on Missing Values

There are several useful methods for detecting, removing, and replacing missing values in Pandas data structures:

- ``isnull()``: generates a boolean mask indicating missing values
- ``notnull()``: generates a boolean mask of non-missing values. Is the opposite of ``isnull()``.
- ``dropna()``: returns a filtered version of the data, without missing values.
- ``fillna()``: returns a copy of the data with missing values filled or imputed with a desired strategy.

Let's review some examples of the first two functions `isnull()` and `notnull()`:

In [ ]:
data = pd.Series([1, np.nan, 'hello', None])
data

0        1
1      NaN
2    hello
3     None
dtype: object

In [ ]:
data.isnull()

0    False
1     True
2    False
3     True
dtype: bool

In [ ]:
data.notnull()

0     True
1    False
2     True
3    False
dtype: bool

As you can see, the `.isnull()` function returns a Dataframe with boolean values, where `False` denotes a present value and `True` denotes a missing value. Conversely, the `.notnull()` function returns a Dataframe with boolean values, where `True` denotes a present value and `False` denotes a missing value.

With this boolean result we can make a subsetting to filter those missing values:

In [ ]:
data[data.notnull()]

0        1
2    hello
dtype: object

In [ ]:
data

0        1
1      NaN
2    hello
3     None
dtype: object

Although we have filtered the missing data, if we review the content of the variable again, we will see that the missing data persists. This is because no subsetting operation is done inplace. Now let's see how we remove missing data from our dataset.

### 5.3. Dropping missing values

The basic function to remove any missing values from a `Series` object is as follows, although the function is not executed inplace:

In [ ]:
data.dropna()

0        1
2    hello
dtype: object

In [ ]:
data

0        1
1      NaN
2    hello
3     None
dtype: object

For a ``DataFrame``, there are more options.
Consider the following ``DataFrame``:

In [ ]:
df = pd.DataFrame([[1,      np.nan, 2],
                   [2,      3,      5],
                   [np.nan, 4,      6]])
df

0    1  2
0  1.0  NaN  2
1  2.0  3.0  5
2  NaN  4.0  6

We cannot drop single values from a ``DataFrame``; we can only drop full rows or full columns.
Depending on the application, you might want one or the other, so ``dropna()`` gives a number of options for a ``DataFrame``.

By default, ``dropna()`` will drop all rows in which *any* missing value is present:

In [ ]:
df.dropna()

0    1  2
1  2.0  3.0  5

Alternatively, you can drop missing values along a different axis; `axis=1` drops all columns containing a missing value:

In [ ]:
df.dropna(axis='columns')

2
0  2
1  5
2  6

But this drops some good data as well; you might rather be interested in dropping rows or columns with *all* `NaN` values, or a majority of `NaN` values. This can be specified through the ``how`` or ``thresh`` parameters, which allow fine control of the number of nulls to allow through.

The default is ``how='any'``, such that any row or column (depending on the ``axis`` keyword) containing a null value will be dropped. You can also specify ``how='all'``, which will only drop rows/columns that are *all* null values:

In [ ]:
df = pd.DataFrame([[1,      np.nan, 2, np.nan],
                   [2,      3,      5, np.nan],
                   [np.nan, 4,      6, np.nan]])
df

0    1  2   3
0  1.0  NaN  2 NaN
1  2.0  3.0  5 NaN
2  NaN  4.0  6 NaN

In [ ]:
df.dropna(axis='columns', how='all')

0    1  2
0  1.0  NaN  2
1  2.0  3.0  5
2  NaN  4.0  6

### 5.4. Filling null values

Sometimes rather than dropping `NaN` values, you'd rather replace them with a valid value.
This value might be a single number like zero, or it might be some sort of imputation or interpolation from the good values.

There are four types of treatment that can be given, in that order, to unwanted non-existent or missing data:

1. **Treatment 1:** Ignore the missing or unwanted data in some columns, considering that in other columns of the same rows there are important or relevant data for the study.
2. **Treatment 2:** Replace the missing or unwanted data with values that represent an indicator of nullity.
3. **Treatment 3:** Replace the missing, nonexistent or unwanted data with interpolated values that are related to the trend of the data that is present.
4. **Treatment 4:** Delete the missing data, with the certainty that valuable information will not be lost when analyzing the data.

You can apply **Treatment 2** and **Treatment 3** in-place using the ``isnull()`` method as a mask, but because it is such a common operation `Pandas` provides the ``fillna()`` method, which returns a copy of the array with the missing values replaced.

Consider the following ``Series``:

In [ ]:
data = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
data

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

To replace the missing values of the Series with a null value, we can do the following:

In [ ]:
data.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

We can specify a forward-fill to propagate the previous value forward:

In [ ]:
# forward-fill
data.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

Or we can specify a back-fill to propagate the next values backward:

In [ ]:
# back-fill
data.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

For ``DataFrame`` objects the options are similar, but we can also specify an ``axis`` along which the fills take place:

In [ ]:
df

0    1  2   3
0  1.0  NaN  2 NaN
1  2.0  3.0  5 NaN
2  NaN  4.0  6 NaN

In [ ]:
df.fillna(method='ffill', axis=1) # forward-fill along the columns

0    1    2    3
0  1.0  1.0  2.0  2.0
1  2.0  3.0  5.0  5.0
2  NaN  4.0  6.0  6.0

In [ ]:
df.fillna(method='ffill', axis=0) # forward-fill along the rows

0    1  2   3
0  1.0  NaN  2 NaN
1  2.0  3.0  5 NaN
2  2.0  4.0  6 NaN

Notice that if a previous value is not available during a forward fill, the `NaN` value remains.

## 6. `Pandas` String Operations

When a `Pandas` object stores string data, `Pandas` provides certain operations to facilitate its manipulation. Let's see what would happen if a classic data storage structure like a list had missing data and a string operation was executed. Firstly, we define a list with four string values:

In [ ]:
data = ['peter', 'Paul', 'MARY', 'gUIDO']
[s.capitalize() for s in data]

['Peter', 'Paul', 'Mary', 'Guido']

This is perhaps sufficient to work with some data, but it will break if there are any missing values:

In [ ]:
data = ['peter', 'Paul', None, 'MARY', 'gUIDO'] #this has a missing value None hence the error below
[s.capitalize() for s in data]

AttributeError: ignored

Now let's look at a Pandas `Series`:

In [ ]:
import pandas as pd
names = pd.Series(data)
names

0    peter
1     Paul
2     None
3     MARY
4    gUIDO
dtype: object

We can now call a single method that will capitalize all the entries, while skipping over any missing values or non-string values:

In [ ]:
names.str.capitalize()

0    Peter
1     Paul
2     None
3     Mary
4    Guido
dtype: object

We have accessed the `str` attribute that parses the values stored in the `Series` to string.

### 6.1. String Methods

Here is a list of Pandas ``str`` methods that mirror Python string methods:

|             |                  |                  |                  |
|-------------|------------------|------------------|------------------|
|``len()``    | ``lower()``      | ``translate()``  | ``islower()``    | 
|``ljust()``  | ``upper()``      | ``startswith()`` | ``isupper()``    | 
|``rjust()``  | ``find()``       | ``endswith()``   | ``isnumeric()``  | 
|``center()`` | ``rfind()``      | ``isalnum()``    | ``isdecimal()``  | 
|``zfill()``  | ``index()``      | ``isalpha()``    | ``split()``      | 
|``strip()``  | ``rindex()``     | ``isdigit()``    | ``rsplit()``     | 
|``rstrip()`` | ``capitalize()`` | ``isspace()``    | ``partition()``  | 
|``lstrip()`` |  ``swapcase()``  |  ``istitle()``   | ``rpartition()`` |

Let's see some examples of string methods for Pandas `Series` with the `monte` Series:

In [ ]:
monte = pd.Series(['Graham Chapman', 'John Cleese', 'Terry Gilliam',
                   'Eric Idle', 'Terry Jones', 'Michael Palin'])

In [ ]:
monte.str.lower() # Parse values to string and transform all characters to lowercase

0    graham chapman
1       john cleese
2     terry gilliam
3         eric idle
4       terry jones
5     michael palin
dtype: object

In [ ]:
monte.str.len() # Parse values to string and calculates their length

0    14
1    11
2    13
3     9
4    11
5    13
dtype: int64

In [ ]:
# Parse values to string and calculates a mask of string values starting by 'T'
monte.str.startswith('T')

0    False
1    False
2     True
3    False
4     True
5    False
dtype: bool

In [ ]:
monte.str.split() # Parse values to string and splits them by ' ' character, by default

0    [Graham, Chapman]
1       [John, Cleese]
2     [Terry, Gilliam]
3         [Eric, Idle]
4       [Terry, Jones]
5     [Michael, Palin]
dtype: object

## Exercise 5

Consider the following lists:

```
country = ['Netherland', 'Germany', 'Peru', 'Israel', 'Madagascar']
year = [2002, 2002, 1957, 2007, 1967]
population = [16122830.0, np.nan, 9146100.0, 6426679.0, 6334556.0]
continent = ['Europe', 'europe', 'Americas', 'asia', 'Africa']
```

1. Create a Dataframe object which contains all the lists values as Series. The final DataFrame should be named as `country_info`, containing 4 columns and 5 rows.
2. Delete the rows which contains missing values
3. Capitalize all the continents in continent column.
4. Get the length of each country's names.

In [ ]:

# Answer 1 : Create DataFrame country_info
# import the libraries
import pandas as pd
import numpy as np

# create list of countries , year , population and continent
country = ['Netherland', 'Germany', 'Peru', 'Israel', 'Madagascar']
year = [2002, 2002, 1957, 2007, 1967]
population = [16122830.0, np.nan, 9146100.0, 6426679.0, 6334556.0]
continent = ['Europe', 'europe', 'Americas', 'asia', 'Africa']

# create dataFrame with the lists
country_info = pd.DataFrame({'country' : country , 'year' : year , 'population' : population , 'continent' : continent})

# view the DataFrame
country_info

country  year  population continent
0  Netherland  2002  16122830.0    Europe
1     Germany  2002         NaN    europe
2        Peru  1957   9146100.0  Americas
3      Israel  2007   6426679.0      asia
4  Madagascar  1967   6334556.0    Africa

In [ ]:
# Answer 2 : Delete rows with missing values
# how many missing values
print(country_info.isnull().sum())

# drop missing values in population axis = rows
country_info.dropna(axis = 0 , inplace = True)

# view the dataframe
country_info

country       0
year          0
population    1
continent     0
dtype: int64


country  year  population continent
0  Netherland  2002  16122830.0    Europe
2        Peru  1957   9146100.0  Americas
3      Israel  2007   6426679.0      asia
4  Madagascar  1967   6334556.0    Africa

In [ ]:
# Answer 3 : Capitalize all the continents in continent column.
country_info['continent'] = country_info['continent'].str.upper()
country_info

country  year  population continent
0  Netherland  2002  16122830.0    EUROPE
2        Peru  1957   9146100.0  AMERICAS
3      Israel  2007   6426679.0      ASIA
4  Madagascar  1967   6334556.0    AFRICA

In [ ]:
# Answer 4 : Get the length of each country's names
# create a coulnn with country_length
country_info['country_length'] = country_info['country'].str.len()
country_info


country  year  population continent  country_length
0  Netherland  2002  16122830.0    EUROPE              10
2        Peru  1957   9146100.0  AMERICAS               4
3      Israel  2007   6426679.0      ASIA               6
4  Madagascar  1967   6334556.0    AFRICA              10

## 7. Concatenate `Series`

Here we'll take a look at simple concatenation of `Series` and `DataFrame` objects with the `pd.concat()` function:

In [ ]:
ser1 = pd.Series(['A', 'B', 'C'], index=[1, 2, 3])
ser2 = pd.Series(['D', 'E', 'F'], index=[4, 5, 6])
pd.concat([ser1, ser2], axis=0)

1    A
2    B
3    C
4    D
5    E
6    F
dtype: object

In [ ]:
pd.concat([ser1, ser2], axis=1)

0    1
1    A  NaN
2    B  NaN
3    C  NaN
4  NaN    D
5  NaN    E
6  NaN    F

By default, the concatenation takes place row-wise within the ``DataFrame`` (i.e., ``axis=0``). However, as you can see, the concatenation of `Series` in a `DataFrame` can be done in contiguous rows or columns by specifying the `axis` parameter. In the case where they are columns, care must be taken to define the same index values, so that the columns are placed contiguously without `NaN` values.

## Exercise 6

Consider the following lists:

```
country = ['Netherland', 'Germany', 'Peru', 'Israel', 'Madagascar']
gdp_per_cap = [33724.757780, 30035.801980, 4245.256698, 25523.277100, 1634.047282]
```

1. Create a Dataframe object which contains all the lists values as Series. The final DataFrame should be named as `country_gdp`, containing 2 columns and 5 rows.
2. Concatenate the two dataframes: `country_info` and `country_gdp` with `axis=0` and name it `concat_data`
3. Check if there are any null values in `concat_data`
4. Find total numer of missing values in each column. *hint: Use `.isnull()` and `.sum()` functions*

In [ ]:
# Answer 1 : Create the DataFrame country_gdp
# create the lists
country = ['Netherland', 'Germany', 'Peru', 'Israel', 'Madagascar']
gdp_per_cap = [33724.757780, 30035.801980, 4245.256698, 25523.277100, 1634.047282]

# create the DataFrame
country_gdp = pd.DataFrame({'country' : country , 'gdp_per_cap' : gdp_per_cap})

# view the data
print(country_gdp)
print(country_info)


      country   gdp_per_cap
0  Netherland  33724.757780
1     Germany  30035.801980
2        Peru   4245.256698
3      Israel  25523.277100
4  Madagascar   1634.047282
      country  year  population continent  country_length
0  Netherland  2002  16122830.0    EUROPE              10
2        Peru  1957   9146100.0  AMERICAS               4
3      Israel  2007   6426679.0      ASIA               6
4  Madagascar  1967   6334556.0    AFRICA              10


In [ ]:
# Answer 2 : Concatenate the two dataframes: country_info and country_gdp with axis=0 and name it concat_data
concat_data = pd.concat([country_gdp , country_info] , axis = 1)
concat_data

country   gdp_per_cap     country  ...  population  continent country_length
0  Netherland  33724.757780  Netherland  ...  16122830.0     EUROPE           10.0
1     Germany  30035.801980         NaN  ...         NaN        NaN            NaN
2        Peru   4245.256698        Peru  ...   9146100.0   AMERICAS            4.0
3      Israel  25523.277100      Israel  ...   6426679.0       ASIA            6.0
4  Madagascar   1634.047282  Madagascar  ...   6334556.0     AFRICA           10.0

[5 rows x 7 columns]

In [ ]:
# Answer 3 : Check if there are any null values in concat_data
concat_data.isnull()

country  gdp_per_cap  country   year  population  continent  country_length
0    False        False    False  False       False      False           False
1    False        False     True   True        True       True            True
2    False        False    False  False       False      False           False
3    False        False    False  False       False      False           False
4    False        False    False  False       False      False           False

In [ ]:
# Answer 4 : Find total numer of missing values in each column.
concat_data.isnull().sum()

country           0
gdp_per_cap       0
country           1
year              1
population        1
continent         1
country_length    1
dtype: int64

## Conclusions

We have learned the fundamentals of the `pandas` library for data analysis, which allows us to execute operations with stored data efficiently, add records and handle missing data. We also highlighted some of the most important functions of Pandas' `Series` and `DataFrame` objects, as well as following a Data Wrangling procedure.

# Practice 
Data source : https://raw.githubusercontent.com/dphi-official/Datasets/master/UNdata_Greenhouse_Gas.csv 
1. Load the CSV file into a variable - greenhouse_data
2. Use the description function to understand how the data looks like
3. Print its first 10 rows using head()
4. Print its last 10 rows using tail()
5. Check if there are any null values
6. Store the DataFrame to a CSV file named ‘file2.csv’

In [ ]:
## QUESTION1 ###
# import the desired libraries
import pandas as pd
import numpy as np

# load the data
greenhouse_data = pd.read_csv('https://raw.githubusercontent.com/dphi-official/Datasets/master/UNdata_Greenhouse_Gas.csv')
print(greenhouse_data.sample(5))

# data types and variable information
print(greenhouse_data.info())

     Country or Area  Year          Value
833           Norway  1996   54291.200895
1168  United Kingdom  1997  747861.872860
1074     Switzerland  2007   52441.788567
898          Romania  2015  116462.878932
200           Cyprus  2013    7843.026427
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1204 entries, 0 to 1203
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  1204 non-null   object 
 1   Year             1204 non-null   int64  
 2   Value            1204 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 28.3+ KB
None


In [ ]:
## QUESTION 2 ##
# convert gas emission to millions
greenhouse_data['Value'] = greenhouse_data['Value']/1000000

# view the data
print(greenhouse_data.sample(5))

# data summary
greenhouse_data.describe()

               Country or Area  Year     Value
155                     Canada  2002  0.724349
1184  United States of America  2009  6.709369
561                      Japan  2016  1.305751
1113                    Turkey  1996  0.267232
127                   Bulgaria  2002  0.059718


Year        Value
count  1204.000000  1204.000000
mean   2003.500000     0.528078
std       8.081104     1.283749
min    1990.000000     0.000084
25%    1996.750000     0.030939
50%    2003.500000     0.080697
75%    2010.250000     0.439248
max    2017.000000     7.369968

1. Data is collected over a period of 27 years (1990 through 2017)  
2. We have three variables i.e. Country-Area , Year and Value of Green House Gas Emission with   
3. There are no missing values 1204 observations

In [ ]:
## QUESTION 3 : FIRST 10 ROWS ##
greenhouse_data.head(10)

Country or Area  Year     Value
0       Australia  2017  0.554127
1       Australia  2016  0.546772
2       Australia  2015  0.535174
3       Australia  2014  0.524957
4       Australia  2013  0.530434
5       Australia  2012  0.540616
6       Australia  2011  0.538281
7       Australia  2010  0.537275
8       Australia  2009  0.540913
9       Australia  2008  0.537032

In [ ]:
## QUESTION 4 : LAST 10 ROWS ##
greenhouse_data.tail(10)

Country or Area  Year     Value
1194  United States of America  1999  7.071461
1195  United States of America  1998  7.032526
1196  United States of America  1997  6.968462
1197  United States of America  1996  6.907699
1198  United States of America  1995  6.710067
1199  United States of America  1994  6.624836
1200  United States of America  1993  6.532070
1201  United States of America  1992  6.424934
1202  United States of America  1991  6.315615
1203  United States of America  1990  6.371001

In [ ]:
## QUESTION 5 : MISSING VALUES ##
greenhouse_data.isnull().sum()

Country or Area    0
Year               0
Value              0
dtype: int64

In [ ]:
## QUESTION 6 : STORE DATAFRAME AS CSV ##
greenhouse_data.to_csv('mygreenhouse.csv')